app/src/main/java/de/westnordost/streetcomplete/quests/address/AddAddressStreet.kt

```diff
@@ -0,0 +1,108 @@
+import de.westnordost.streetcomplete.data.osm.ElementGeometry
+import de.westnordost.streetcomplete.data.osm.ElementPolylinesGeometry
+import de.westnordost.streetcomplete.data.osm.tql.getQuestPrintStatement
+import de.westnordost.streetcomplete.data.osm.tql.toGlobalOverpassBBox
+import de.westnordost.streetcomplete.quests.localized_name.data.RoadNameSuggestionsDao
+import java.util.regex.Pattern
+
+class AddAddressStreet(
+        private val overpassServer: OverpassMapDataAndGeometryDao,
+        private val roadNameSuggestionsDao: RoadNameSuggestionsDao
+) : OsmElementQuestType<AddressStreetAnswer> {
+    override val commitMessage = "Add street/place names to address"
+    override val icon = R.drawable.ic_quest_housenumber_street
+
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_address_street_title
+
+    override fun getTitleArgs(tags: Map<String, String>, featureName: Lazy<String?>): Array<String> {
+        val housenumber = tags["addr:housenumber"]
+        return if (housenumber != null) arrayOf(housenumber) else arrayOf()
+    }
+
+    override fun createForm() = AddAddressStreetForm()
+
+    override fun isApplicableTo(element: Element): Boolean? = null
+
+    override fun download(bbox: BoundingBox, handler: (element: Element, geometry: ElementGeometry?) -> Unit): Boolean {
+        Log.wtf("aaa", getOverpassQuery(bbox))
+        if (!overpassServer.query(getOverpassQuery(bbox), handler)) return false
+        if (!overpassServer.query(getStreetNameSuggestionsOverpassQuery(bbox), this::putRoadNameSuggestion)) return false
+        return true
+    }
+
+    private fun getOverpassQuery(bbox: BoundingBox) =
+            bbox.toGlobalOverpassBBox() + """
+            relation["type"="associatedStreet"];
+            > -> .inStreetRelation;
+
+            way["highway" ~ "^(${OsmTaggings.ALL_ROADS.joinToString("|")})$"]["name"] -> .named_roads;
+
+            nwr["addr:street"!~".*"]["addr:housenumber"]["addr:place"!~".*"](around.named_roads:150) -> .missing_data;
+
+            (.missing_data; - .inStreetRelation;);""" +
+                    getQuestPrintStatement()
+
+    // TODO: probably either this should be modified to match AddRoadName with custom set of roads  with config array migrated to OsmTaggings
+    // TODO: or AddRoadName can use something like that
+    /** return overpass query string to get roads with names
+     * */
+    private fun getStreetNameSuggestionsOverpassQuery(bbox: BoundingBox) =
+            bbox.toGlobalOverpassBBox() + "\n" + """
+        way[highway ~ "^(${OsmTaggings.ALL_ROADS.joinToString("|")})$"][name];
+        out body geom;""".trimIndent()
+
+    override fun applyAnswerTo(answer: AddressStreetAnswer, changes: StringMapChangesBuilder) {
+        var key = ""
+        when(answer){
+            is StreetName -> {key = "addr:street"}
+            is PlaceName -> {key = "addr:place"}
+        }
```

Comment:
You can shorten that to

```suggestion
        val key = when(answer) {
            is StreetName -> "addr:street"
            is PlaceName -> "addr:place"
        }
```